base/ltcmd.dtx

```diff
@@ -3073,12 +3263,48 @@
 %   prevents loss of braces, and there is then a test to see if there are
 %   nested delimiters to handle.
 %    \begin{macrocode}
-\cs_new_protected:Npn \@@_grab_D_aux:NNnNNN #1#2#3#4#5#6
+\group_begin:
+  \char_set_catcode_other:N \^^M
+  \cs_new_protected:Npn \@@_grab_D_aux:NNnNNNN #1#2#3#4#5#6#7
+    {
+      \@@_grab_D_aux:NNnNN #1#2 {#3} #4 #6
+      #7
+        {
+          \group_begin:
+            \@@_grab_D_verb_safe:N #1
+        }
+      #5 #1
+        {
+          #7 { \group_end: }
+          \@@_grab_D_call:Nw #1
+        }
+        {
+          #7 { \group_end: }
+          \@@_add_arg:o \c_novalue_tl
+        }
+    }
+\group_end:
+%    \end{macrocode}
+%   The only awkwardness here is the need to preserve the catcode of the search
+%   token: this is done low-level for performance reasons. Only values that
+%   are realistic are included.
+%    \begin{macrocode}
+\cs_new_protected:Npn \@@_grab_D_verb_safe:N #1
   {
-    \@@_grab_D_aux:NNnNN #1#2 {#3} #4 #6
-    #5 #1
-      { \@@_grab_D_call:Nw #1 }
-      { \@@_add_arg:o \c_novalue_tl }
+    \cs_set_eq:NN \do \char_set_catcode_other:N
+    \dospecials
+    \char_set_catcode_other:N \^^M
+    \use:c
+      {
+        char_set_catcode_
+        \if_catcode:w \c_math_toggle_token #1 math_toggle \else:
+        \if_catcode:w ^ #1 math_subscript \else:
+        \if_catcode:w \c_math_subscript_token #1 math_superscript \else:
+        \if_catcode:w A #1 letter \else:
+        other \fi: \fi: \fi: \fi:
+        :N
+      }
+        #1
```

Comment:
I *think* we need this treatment for both the start and end token of the delimited optional argument, don't we?

Reply:
No. For the opening token we are using `\futurelet`, so tokenization is an issue. Once we have the opening token, we _know_ there should be a closing one - that is therefore handled using a delimited argument. The closing token therefore never looks ahead, and we are safe leaving catcodes alone.